In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [6]:
df = pd.read_csv('dulieuxettuyendaihoc.csv',
                 header=0,delimiter=',',encoding='utf-8')

df.rename(columns={'NGAYTHI':'NT',
                   'DINHHUONGNGHENGHIEP':'NGHE'}, 
                   inplace=True)
df

MSSV GT  DT   KV    M1    M2    M3  KT         NT NGHE
0     SV001  F NaN  2NT  3.25  3.25  4.50  A1  12/7/2018   No
1     SV002  M NaN    1  6.00  4.00  3.50   C  12/7/2018  Yes
2     SV003  M NaN    1  5.00  6.75  4.00   C  12/7/2018   No
3     SV004  M NaN    1  4.25  4.25  5.25  D1  12/7/2018   No
4     SV005  M NaN  2NT  4.25  4.50  5.00   A  12/7/2018   No
..      ... ..  ..  ...   ...   ...   ...  ..        ...  ...
95   SV0096  F NaN    1  5.25  1.50  6.25   C  7/20/2022  Yes
96   SV0097  F NaN    1  5.25  3.75  4.75   C  7/20/2022   No
97   SV0098  M NaN  2NT  7.00  8.00  4.00   C  7/20/2022  Yes
98   SV0099  M NaN    1  5.00  3.50  5.50   C  7/20/2022   No
99  SV00100  M NaN  2NT  5.25  2.50  4.25   C  7/20/2022  Yes

[100 rows x 10 columns]

### PP1: One Sample T-Test (so luong mau < 30): bien định lượng
### H0: muy = x


In [3]:
dfKhoiC = df.loc[df['KT'] == 'C']
dfKhoiC['M1']

1     6.00
2     5.00
6     6.50
22    5.00
23    6.75
24    7.00
25    4.75
26    5.25
27    5.25
95    5.25
96    5.25
97    7.00
98    5.00
99    5.25
Name: M1, dtype: float64

### giả định: H0: điểm trung bình M1 = 4.0

In [4]:
import scipy.stats as stats

stats.ttest_1samp(dfKhoiC['M1'], popmean=4.0)

TtestResult(statistic=np.float64(7.650519779016341), pvalue=np.float64(3.6339826927369557e-06), df=np.int64(13))

### pvalue = 3.6339826927369557e-06 < alpha = 5%
### => bác bỏ giả định: ko đủ dữ kiện để nói rằng tb điểm thi M1 của những thí sinh thi khối C là 4.0 với  mức sai lầm là 5%

# kiểm định tb điểm M2 của thí sinh thi khối C có bằng 5.7 ko

In [5]:
dfKhoiC = df.loc[df['KT'] == 'C']
dfKhoiC['M2']
stats.ttest_1samp(dfKhoiC['M2'], popmean=5.7)

TtestResult(statistic=np.float64(-2.820074837915494), pvalue=np.float64(0.0144629751431775), df=np.int64(13))

### pvalue = 0.01 < alpha = 5%
### => bác bỏ giả định: ko đủ dữ kiện để nói rằng tb điểm thi M2 của những thí sinh thi khối C là 5.7 với  mức sai lầm là 5%

# PP2: Two-sample T-Test( n < 30): biến định lượng
# Giả định H0: Trung bình X = Trung bình Y

#### hãy kiểm tra xem tb điểm thi M1 và trung bình điểm thi M2 của các thí sinh thi khối C có bằng nhau ko

In [7]:
dfKhoiC = df.loc[df['KT'] == 'C']
stats.ttest_ind(dfKhoiC['M1'], dfKhoiC['M2'], equal_var=True)

TtestResult(statistic=np.float64(2.5331368604654054), pvalue=np.float64(0.017676922552296807), df=np.float64(26.0))

#### pvalue = 0.0176 < alpha = 0.05
#### => bác bỏ H0: ko đủ dữ kiện đê nói rằng tb điểm thi M1 và trung bình điểm thi M2 của các thí sinh thi khối C bằng nhau

## suy diễn trung bình diểm M1 và trung bình điểm M3 của thí sinih thi khối C có bằng nhau ko

In [8]:
dfKhoiC = df.loc[df['KT'] == 'C']
stats.ttest_ind(dfKhoiC['M1'], dfKhoiC['M3'], equal_var=True)

TtestResult(statistic=np.float64(2.6321295849540447), pvalue=np.float64(0.014085909192033959), df=np.float64(26.0))

### pvl = 0.014 < 0.05
### ko đủ dữ kiện đê nói rằng tb điểm thi M1 và trung bình điểm thi M3 của các thí sinh thi khối C bằng nhau

# PP3: one-smaple Z-test (n >= 30) H0= x

#### hảy suy điển xem tb điểm thi M1 có bằng 8.0 ko

In [9]:
from statsmodels.stats.weightstats import ztest as ztest

ztest(df['M1'], value=8.0)

(np.float64(-29.907152266677993), np.float64(1.588509177330612e-196))

### pvl= 1.588509177330612e> 0.05 => bác bỏ

In [10]:
ztest(df['M1'], value=6.7)

(np.float64(-20.78055650454621), np.float64(6.49019792505497e-96))

# PP4: Two -smaple z-test  H0: trung bình X = trung bình y

In [11]:
ztest(df['M1'], df['M2'], value=0)

(np.float64(-2.9004757923795), np.float64(0.0037259661678783573))

p = 0.003 < 0.05 => bác bỏ: H0: ko đủ đk kết luận răng tb M1 bằng tb M2

In [12]:
ztest(df['M1'], df['M3'], value=0)

(np.float64(-6.432991757573295), np.float64(1.2511643846506845e-10))

p = 1.2511643846506845e-10 < 0.05
=> bác bỏ H0 => ko đủ đk kết luận răng tb M1 bằng tb M3

# PP5: suy diễn trương quan: 2 biến định lượng (X,Y)
### hệ số tương quan (r):
### r<0: tương quan nghịch 
### r=0: ko tương quan
### r>0: tương quan thuận 
### giá trị tương quan (r) nằm [-1;1]
### |r| là dộ lớn 
### < 0.5 là yếu
### 0.5 < r <0.65: trung bình
### 0.65 < r< 0.8: khá
### 0.8 < r < 0.95: tốt
### > 0.95: hoàn hảo
## H0: r  = 0

In [13]:
from scipy.stats.stats import pearsonr

C:\Users\Admin\AppData\Local\Temp\ipykernel_6568\4005289447.py:1: DeprecationWarning: Please import `pearsonr` from the `scipy.stats` namespace; the `scipy.stats.stats` namespace is deprecated and will be removed in SciPy 2.0.0.
  from scipy.stats.stats import pearsonr


In [14]:
pearsonr(df['M1'], df['M2'])

PearsonRResult(statistic=np.float64(0.03515405725785023), pvalue=np.float64(0.7284198928723625))

p = 0.73 > 0.05=> chấp nhận H0: ko đủ đk để nói rằng M1 và M2 tương quan hay nói cách khác đủ đk để nói rằng M1 và M2 ko tương quan
H0 là r = 0 mà p > alpha nên chấp nhận H0 => r = 0 là ko tương quan 

In [15]:
pearsonr(df['M1'], df['M3'])

PearsonRResult(statistic=np.float64(0.018730760164383135), pvalue=np.float64(0.8532534347055037))

p = 0.85 > 0.05=> chấp nhận H0: ko đủ đk để nói rằng M1 và M2 tương quan hay nói cách khác đủ đk để nói rằng M1 và M3 ko tương quan
H0 là r = 0 mà p > alpha nên chấp nhận H0 => r = 0 là ko tương quan 

In [16]:
pearsonr(df['M2'], df['M3'])

PearsonRResult(statistic=np.float64(-0.09152288755408225), pvalue=np.float64(0.3651344767496172))

p = 0.36 > 0.05=> chấp nhận H0: ko đủ đk để nói rằng M2 và M3 tương quan hay nói cách khác đủ đk để nói rằng M1 và M3 ko tương quan
H0 là r = 0 mà p > alpha nên chấp nhận H0 => r = 0 là ko tương quan 

# PP6: Suy diễn Fisher: 2 biến định tính (giá trị nhị phân 0 và 1)
### H0: 2 biến độc lập


In [17]:
import scipy.stats as stats

kiểm tra có sự phụ thuộc nào giữa việc sịnh viên có định hướng nghề nghiệp và giới tinh ko

In [20]:
#khi dùng pp định tính này cần lặp bảng này
crosdata = pd.crosstab(df['GT'], [df['NGHE']], rownames=['GT'], colnames=['NGHE'])
crosdata

NGHE  No  Yes
GT           
F     23   25
M     32   20

yes/no (0,1) và f/m(0,1) là số nhị phân nên sd pp này

# H0: NN độc lập với GT

In [21]:
odd_ratio, p_value = stats.fisher_exact(crosdata)
print('odd ratio is : ' + str(odd_ratio))
print('p_value is : ' + str(p_value))


odd ratio is : 0.575
p_value is : 0.22763927303454412


p = 0.227 > alpha = 5%
=> chấp nhận H0 => đủ đk để nói rang NN độc lập GT hoặc ko đủ đk để nói rằng NN phụ thuộc GT 

# PP7: Suy diễn ChiSquare: 2 biến đinh tính (gtri ko phải nhị phân)
# H0 là 2 biến độc lập

In [22]:

from scipy.stats import chi2_contingency

crosdata = pd.crosstab(df['KV'], [df['KT']], rownames=['KV'], colnames=['KT'])
crosdata

KT    A  A1  B  C  D1
KV                   
1    29   2  8  8  13
2     9   0  0  2   8
2NT  11   4  1  4   1

In [24]:
stat, p, dof, expected = chi2_contingency(crosdata)
 
alpha = 0.05
print("p value is " + str(p))
if p <= alpha:
    print('Dependent (reject H0)')
else:
    print('Independent (H0 holds true)')

p value is 0.02012461887796485
Dependent (reject H0)


ko đủ đk để nói rằng KT ko phụ thuộc vào KV

In [26]:
crosdata = pd.crosstab(df['GT'], [df['KT']], rownames=['GT'], colnames=['KT'])
crosdata

KT   A  A1  B  C  D1
GT                  
F   15   5  4  8  16
M   34   1  5  6   6

In [27]:
stat, p, dof, expected = chi2_contingency(crosdata)
 
alpha = 0.05
print("p value is " + str(p))
if p <= alpha:
    print('Dependent (reject H0)')
else:
    print('Independent (H0 holds true)')

p value is 0.005044752209452435
Dependent (reject H0)


GT phụ thuộc vào KT, ko đủ đk để nói rằng KT độc lập vào GT

# 8. Suy diễn One way Anova: định lượng và định tính
# giả định: H0: trung bình biến định lượng trên từng nhóm biến định tính là bằng nhau

#### điem M1 có phụ thuộc vào Gt ko
H0: trung bình M1 trên nhóm Female = tb M1 trên nhóm male

In [28]:

import statsmodels.api as sm
from statsmodels.formula.api import ols

model = ols('M1 ~ GT', data=df).fit()
aov_table = sm.stats.anova_lm(model, typ=1)
aov_table

df      sum_sq   mean_sq         F    PR(>F)
GT         1.0    1.203141  1.203141  0.590538  0.444059
Residual  98.0  199.661859  2.037366       NaN       NaN

p = 0.444059 > 0.05 => chấp nhận => trung bình M1 trên nhóm Female = tb M1 trên nhóm male => M1 ko phụ thuộc vào GT hay nói cách khác điểm tb M1 độc lập vs GT

In [ ]:
# M2 có phụ thuộc vào KV hay ko
# M3 có phụ thuoc vao KT ko

In [29]:
model = ols('M2 ~ KV', data=df).fit()
aov_table = sm.stats.anova_lm(model, typ=1)
aov_table

df      sum_sq   mean_sq         F    PR(>F)
KV         2.0    6.053398  3.026699  2.790934  0.066299
Residual  97.0  105.194102  1.084475       NaN       NaN

In [31]:
model = ols('M3~ KT', data=df).fit()
aov_table = sm.stats.anova_lm(model, typ=1)
aov_table

df      sum_sq   mean_sq         F   PR(>F)
KT         4.0    3.967636  0.991909  0.911814  0.46041
Residual  95.0  103.344864  1.087841       NaN      NaN

# 9. Two way ANOVA
# p > alpha là dộc lap, nhỏ hơn là phụ thuoc

In [32]:
model = ols('M1 ~ GT + KV + GT:KV',data=df).fit()
result = sm.stats.anova_lm(model, type=2)
print(result)

            df      sum_sq   mean_sq         F    PR(>F)
GT         1.0    1.203141  1.203141  0.598960  0.440917
KV         2.0    3.493206  1.746603  0.869512  0.422500
GT:KV      2.0    7.349290  3.674645  1.829350  0.166186
Residual  94.0  188.819363  2.008717       NaN       NaN


#### pGT = 0.440917 độc lap
####  pKV = 0.422500 độc lap
 ####  pGT:KV = 0.166186 độc lap

In [33]:
model = ols('M2 ~ KV + KT + KV:KT',data=df).fit()
result = sm.stats.anova_lm(model, type=2)
print(result)

            df     sum_sq   mean_sq         F    PR(>F)
KV         2.0   6.053398  3.026699  2.987203  0.055626
KT         4.0  10.061984  2.515496  2.482671  0.049539
KV:KT      8.0   8.688324  1.086040  1.071869  0.390293
Residual  87.0  88.150282  1.013222       NaN       NaN


##### M2 và KV: độc lập
##### M2 và KT: ko độc lập
##### M2 trên từng nhóm KV theo từng loại KT: độc lập

# 10. Phân tích hồi quy
dùng để phân tích sự ảnh hưởng của 1 hay nhiều biến định lượng đến 1 biến định lượng

VD: hãy cho biến sự ảnh hưởng của M1 đến M3


M3 = f(M1)

M1 là biến độc lập

M3 là biến mục tiêu

#### hàm hồi quy tuyến tính
#### A0 + A1 * M1
#### A0: hệ số tự do


In [34]:
import statsmodels.api as sm
X_with_constant = sm.add_constant(df[["M1"]].values)
y = df[['M3']].values
result = sm.OLS(y, X_with_constant).fit()
print(result.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                 -0.010
Method:                 Least Squares   F-statistic:                   0.03439
Date:                Mon, 09 Sep 2024   Prob (F-statistic):              0.853
Time:                        19:55:27   Log-Likelihood:                -145.41
No. Observations:                 100   AIC:                             294.8
Df Residuals:                      98   BIC:                             300.0
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          4.8238      0.295     16.338      0.0

R-squared = 0.000 (*100% = 0% => mô hình ko có ý nghĩa)

 Prob (F-statistic) = 0.853 > 5% => mô hình k có ý nghĩa ( < 5% thì có nghĩa)

const (A0 ) = 4.8238 

x1 (A1) = 0.0137 

=>phương trình:  M3 = 4.8238  + 0.0137  * M1

nếu M1 = 6.7

=>M3 = 4.8238  + 0.0137 * 6.7 = 4.91

# nếu M2 = 5.7 thì M3 bằng bao nhiu

In [36]:
import statsmodels.api as sm
X_with_constant = sm.add_constant(df[["M2"]].values)
y = df[['M3']].values
result = sm.OLS(y, X_with_constant).fit()
print(result.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.008
Model:                            OLS   Adj. R-squared:                 -0.002
Method:                 Least Squares   F-statistic:                    0.8278
Date:                Mon, 09 Sep 2024   Prob (F-statistic):              0.365
Time:                        20:06:06   Log-Likelihood:                -145.00
No. Observations:                 100   AIC:                             294.0
Df Residuals:                      98   BIC:                             299.2
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          5.2575      0.433     12.139      0.0

In [37]:
5.2575 - 0.0899  * 5.7


4.74507

# Phân tích sự ảnh hưởng M1,M2 đến M3

# M3 = f(M1,M2)
A0 + A1 * M1 + A2 * M2
# M1 = 6.0 
# M2 = 5.5

In [38]:
import statsmodels.api as sm
X_with_constant = sm.add_constant(df[["M1","M2"]].values)
y = df[['M3']].values
result = sm.OLS(y, X_with_constant).fit()
print(result.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.009
Model:                            OLS   Adj. R-squared:                 -0.012
Method:                 Least Squares   F-statistic:                    0.4335
Date:                Mon, 09 Sep 2024   Prob (F-statistic):              0.649
Time:                        20:08:54   Log-Likelihood:                -144.98
No. Observations:                 100   AIC:                             296.0
Df Residuals:                      97   BIC:                             303.8
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          5.2006      0.508     10.241      0.0

In [39]:
M3 = 5.2006 +  0.0161 * 6 -0.0906 * 5.5
M3

4.798899999999999

In [40]:
X = df[["M1","M2"]].values
y = df[['M3']].values
result = sm.OLS(y, X).fit()
print(result.summary())

                                 OLS Regression Results                                
Dep. Variable:                      y   R-squared (uncentered):                   0.911
Model:                            OLS   Adj. R-squared (uncentered):              0.909
Method:                 Least Squares   F-statistic:                              500.8
Date:                Mon, 09 Sep 2024   Prob (F-statistic):                    3.54e-52
Time:                        20:14:05   Log-Likelihood:                         -181.63
No. Observations:                 100   AIC:                                      367.3
Df Residuals:                      98   BIC:                                      372.5
Df Model:                           2                                                  
Covariance Type:            nonrobust                                                  
                 coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------

|M1| = 0.4062   < |M2| = 0.7367 => M2 tác động đến M3 hơn